### PR TITLE
Make Tree type abstract, add views

### LINKED ISSUE #2: abstraction & views for pattern matching
If we made the types abstract and only exported smart constructors, we'd gain the benefit of being able to change the underlying representation without breaking client code. However, it would make the library a bit less nice to use, since pattern matching wouldn't be available.

What I usually do in my ML code is write a "view" type which lets me inject or project one layer of structure in a pattern-matchable way. Here's an example:

``` purescript
data Tree a -- abstract
data View f a = Tree a [f a]

inject :: View Tree a -> Tree a
project :: Tree a -> View Tree a
```

Then, users can use the `inject,project` gadgets to build and deconstruct trees, with pattern matching:

``` purescript
cons :: a -> [Tree a] -> Tree a
cons x ts = project $ Tree x ts

head :: Tree a -> a
head t = 
  case project t of
    Tree x xs -> x
```

No need to implement this right away; but eventually, I think this is a pretty nice/painless approach that balances the two issues (abstraction vs. ease of use).

### PR DESCRIPTION
**Pull request by jonsterling**
Resolves #2.

This is an example of how to do it; it's not imperative that this be merged immediately, but I thought I would put it up as a proof of concept. The approach of using a pattern functor is pretty standard in ML, as well as in certain regional dialects of Haskell... I have found in practice after using this technique quite extensively that it balances out the (usually) competing requirements of modularity and ease of use pretty nicely.

One of the nice things about this is that we could provide multiple view types (for instance, one that used lists and one that used arrays, or something even more elaborate). We can change the underlying tree representation at will, and only incur the cost of conversion at a single layer of structure in the view.

Also, feel free to bikeshed on names (including `View`, `into`, `out`, and `transView`). Another option might be to call it a `Pattern`... I don't personally have any preference.

### CODE REVIEW

File: src/Data/Tree.purs [hunk 1]
```diff
@@ -9,8 +13,35 @@ import Data.Foldable (Foldable, foldr, foldl, foldMap)
...
+-- |
+-- | ````purescript
+-- | tree :: Tree Int
+-- | tree = into $ View 0 []
+-- | ````
+into :: forall a. View Tree a -> Tree a
+into (View a ts) = Tree a ts
+
+-- | Map a tree across a natural transformation.
+transView :: forall f g a. (forall a. f a -> g a) -> View f a -> View g a
```
Comment (puffnfresh) [inline src/Data/Tree.purs]: Interesting stuff!

Is this going to give a warning in the future? https://github.com/purescript/purescript/issues/1465

Comment (jonsterling) [inline src/Data/Tree.purs]: @puffnfresh thank you for the advance warning! I'll change the variable name...

**Comment by jonsterling** [comment]
I think that this PR is probably ready—@matthieubulte @jplatte Would you mind reviewing & merging if it is acceptable? Or, I'm happy to make any modification that sounds reasonable. This way, I can proceed with the work on `Data.Tree.Fan` and `Data.Tree.Spread`, which is currently based atop this branch.

By the way, a more elaborate example of views (i.e. where we vary the `f` parameter) can be seen in my branch where I'm working on non-wellfounded trees: https://github.com/jplatte/purescript-trees/blob/db90071629470916c49d3259086f52c20ac57945/docs/Data/Fan.md

**Comment by matthieubulte** [comment]
As I said in Gitter, that looks quite good so go ahead and merge it :+1:

**Comment by jonsterling** [comment]
@matthieubulte Cool, thanks! (Sorry for the redundancy, just wanted to make sure that anyone who hadn't seen Gitter had a chance to comment.)

**Pull request merged**